src/Middleware/ConcurrencyLimiter/test/PolicyTests/ResettableBooleanCompletionSourceTests.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.AspNetCore.ConcurrencyLimiter.Tests.PolicyTests
+{
+    public static class ResettableBooleanCompletionSourceTests
+     {
+        private static StackPolicy _testQueue = TestUtils.CreateStackPolicy(8);
+
+        [Fact]
+        public async static void CanBeAwaitedMultipleTimes()
+        {
+            var tcs = new ResettableBooleanCompletionSource(_testQueue);
+
+            tcs.Complete(true);
+            Assert.True(await tcs.GetValueTask());
+
+            tcs.Complete(true);
+            Assert.True(await tcs.GetValueTask());
+
+            tcs.Complete(false);
+            Assert.False(await tcs.GetValueTask());
+
+            tcs.Complete(false);
+            Assert.False(await tcs.GetValueTask());
+        }
+
+        [Fact]
+        public async static void CanSetResultToTrue()
+        {
+            var tcs = new ResettableBooleanCompletionSource(_testQueue);
+
+            _ = Task.Run(() =>
+            {
+                tcs.Complete(true);
+            });
+
+            var result = await tcs.GetValueTask();
+            Assert.True(result);
+        }
+
+        [Fact]
+        public async static void CanSetResultToFalse()
+        {
+            var tcs = new ResettableBooleanCompletionSource(_testQueue);
+
+            _ = Task.Run(() =>
+            {
+                tcs.Complete(false);
+            });
+
+            var result = await tcs.GetValueTask();
+            Assert.False(result);
+        }
+
+        [Fact]
+        public static void DoubleCallToGetResultCausesError()
+        {
+            // important to verify it throws rather than acting like a new task
+
+            var tcs = new ResettableBooleanCompletionSource(_testQueue);
+            var task = tcs.GetValueTask();
+            tcs.Complete(true);
+
+            Assert.True(task.Result);
+            Assert.Throws<InvalidOperationException>(() => task.Result);
+        }
+
+        [Fact]
+        public static Task RunsContinuationsAsynchronously()
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            async void RunTest()
+            {
+                try
+                {
+                    await RunsContinuationsAsynchronouslyInternally();
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
+                    throw;
+                }
+
+                tcs.SetResult(null);
+            }
+
+            // The Xunit TestSyncContext causes the resettable tcs to always dispatch in effect.
+            ThreadPool.UnsafeQueueUserWorkItem(_ => RunTest(), state: null);
+
+            return tcs.Task;
+        }
+
+        private static async Task RunsContinuationsAsynchronouslyInternally()
+        {
+            var tcs = new ResettableBooleanCompletionSource(_testQueue);
+            var mre = new ManualResetEventSlim();
```

Comment:
nit: can this test work with a TaskCompletionSource instead of MRE.

Reply:
You could use a TCS, but you'd have to call .Wait() or something on it. It probably is a good idea to wrap the test in a try/finally, and call .Set() on the MRE in the finally so this test doesn't end up blocking a thread indefinitely if it fails.